### ISSUE TITLE
Setting LC_NUMERIC breaks parsing of constants

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13459]
Migrated from [rt.perl.org#120723](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=120723) (status was 'resolved')

Searchable as RT120723$

**Comment by p5pRT** [comment]
### From kaffeetisch@gmx.de

#### Created by kaffeetisch@gmx.de



It seems like setting LC\_NUMERIC to any locale with a decimal separator
distinct from the dot breaks parsing of constants&#8203;:

\# perl \-e'use POSIX qw/locale\_h/; BEGIN\{ setlocale \(LC\_NUMERIC\,
"de\_DE\.UTF\-8"\); \} use 5\.008;'
Invalid version format \(non\-numeric data\) at \-e line 1\.
BEGIN failed\-\-compilation aborted at \-e line 1\.

Porting/bisect\.pl puts the blame on this commit&#8203;:

commit 01ec34b5d8c7075801bdb8ca8364cbe2027147dd
Author&#8203;: Karl Williamson \<public@&#8203;khwilliamson\.com>
Date&#8203;:   Thu Jun 20 12&#8203;:28&#8203;:12 2013 \-0600

&nbsp;    util\.c&#8203;: Avoid unnecessary setlocale\(\) calls

&nbsp;    This code sets the locale to C around its work\.  This is unnecessary if
&nbsp;    the locale is already C\, and there is an existing global that indicates
&nbsp;    this\, that can be tested to avoid the sets\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.19.6:

Configured by torsten at Sat Dec  7 17:48:52 CET 2013.

Summary of my perl5 (revision 5 version 19 subversion 6) configuration:

  Platform:
    osname=linux, osvers=3.8.0-34-generic,
archname=x86_64-linux-thread-multi
    uname='linux home.hcsiteeffak.de 3.8.0-34-generic #49-ubuntu smp tue
nov 12 18:00:10 utc 2013 x86_64 x86_64 x86_64 gnulinux '
    config_args='-de -Dprefix=/opt/perlbrew/perls/perl-5.19.6
-Dusethreads -DDEBUGGING -Dusedevel
-Aeval:scriptdir=/opt/perlbrew/perls/perl-5.19.6/bin'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBUGGING
-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include
-D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -g',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBUGGING
-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.7.3', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib/x86_64-linux-gnu /lib/../lib
/usr/lib/x86_64-linux-gnu /usr/lib/../lib /lib /usr/lib
    libs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.17'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -g -L/usr/local/lib
-fstack-protector'



@INC for perl 5.19.6:

/opt/perlbrew/perls/perl-5.19.6/lib/site_perl/5.19.6/x86_64-linux-thread-multi
    /opt/perlbrew/perls/perl-5.19.6/lib/site_perl/5.19.6
    /opt/perlbrew/perls/perl-5.19.6/lib/5.19.6/x86_64-linux-thread-multi
    /opt/perlbrew/perls/perl-5.19.6/lib/5.19.6
    .


Environment for perl 5.19.6:
    HOME=/home/torsten
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LC_ADDRESS=de_DE.UTF-8
    LC_IDENTIFICATION=de_DE.UTF-8
    LC_MEASUREMENT=de_DE.UTF-8
    LC_MONETARY=de_DE.UTF-8
    LC_NAME=de_DE.UTF-8
    LC_NUMERIC=de_DE.UTF-8
    LC_PAPER=de_DE.UTF-8
    LC_TELEPHONE=de_DE.UTF-8
    LC_TIME=de_DE.UTF-8
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)

PATH=/opt/perlbrew/bin:/opt/perlbrew/perls/perl-5.19.6/bin:/opt/bin:/home/torsten/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games
    PERLBREW_BASHRC_VERSION=0.65
    PERLBREW_HOME=/home/torsten/.perlbrew
    PERLBREW_MANPATH=/opt/perlbrew/perls/perl-5.19.6/man
    PERLBREW_PATH=/opt/perlbrew/bin:/opt/perlbrew/perls/perl-5.19.6/bin
    PERLBREW_PERL=perl-5.19.6
    PERLBREW_ROOT=/opt/perlbrew
    PERLBREW_VERSION=0.65
    PERL_BADLANG (unset)
    SHELL=/usr/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From @khwilliamson

On 12/07/2013 02&#8203;:56 PM\, Torsten Schoenfeld \(via RT\) wrote&#8203;:
> \# New Ticket Created by  Torsten Schoenfeld
> \# Please include the string&#8203;:  \[perl \#120723\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=120723 >
>
>
> This is a bug report for perl from kaffeetisch@&#8203;gmx\.de\,
> generated with the help of perlbug 1\.39 running under perl 5\.19\.6\.
>
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
>
> It seems like setting LC\_NUMERIC to any locale with a decimal separator
> distinct from the dot breaks parsing of constants&#8203;:
>
> \# perl \-e'use POSIX qw/locale\_h/; BEGIN\{ setlocale \(LC\_NUMERIC\,
> "de\_DE\.UTF\-8"\); \} use 5\.008;'
> Invalid version format \(non\-numeric data\) at \-e line 1\.
> BEGIN failed\-\-compilation aborted at \-e line 1\.
>
> Porting/bisect\.pl puts the blame on this commit&#8203;:
>
> commit 01ec34b5d8c7075801bdb8ca8364cbe2027147dd
> Author&#8203;: Karl Williamson \<public@&#8203;khwilliamson\.com>
> Date&#8203;:   Thu Jun 20 12&#8203;:28&#8203;:12 2013 \-0600
>
>      util\.c&#8203;: Avoid unnecessary setlocale\(\) calls
>
>      This code sets the locale to C around its work\.  This is unnecessary if
>      the locale is already C\, and there is an existing global that indicates
>      this\, that can be tested to avoid the sets\.
>
> \[Please do not change anything below this line\]
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \-\-\-
> Flags&#8203;:
>      category=core
>      severity=low
> \-\-\-
> Site configuration information for perl 5\.19\.6&#8203;:
>
> Configured by torsten at Sat Dec  7 17&#8203;:48&#8203;:52 CET 2013\.
>
> Summary of my perl5 \(revision 5 version 19 subversion 6\) configuration&#8203;:
>
> \-\-\-
> Environment for perl 5\.19\.6&#8203;:
>      HOME=/home/torsten
>      LANG=en\_US\.UTF\-8
>      LANGUAGE \(unset\)
>      LC\_ADDRESS=de\_DE\.UTF\-8
>      LC\_IDENTIFICATION=de\_DE\.UTF\-8
>      LC\_MEASUREMENT=de\_DE\.UTF\-8
>      LC\_MONETARY=de\_DE\.UTF\-8
>      LC\_NAME=de\_DE\.UTF\-8
>      LC\_NUMERIC=de\_DE\.UTF\-8
>      LC\_PAPER=de\_DE\.UTF\-8
>      LC\_TELEPHONE=de\_DE\.UTF\-8
>      LC\_TIME=de\_DE\.UTF\-8
>      LD\_LIBRARY\_PATH \(unset\)
>      LOGDIR \(unset\)
>

I couldn't reproduce this until I looked at the startup environment that 
perlbug graciously furnishes\.  It's the  LC\_NUMERIC=de\_DE\.UTF\-8
that causes the problem\.  I will work on a patch\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @khwilliamson

Fixed by bc8ec7cc020d0562094a551b280fd3f32bf5eb04
\-\- 
Karl Williamson

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'open' to 'resolved'
